src/dbnode/storage/series/types.go

```diff
@@ -283,3 +284,35 @@ func NewStats(scope tally.Scope) Stats {
 func (s Stats) IncCreatedEncoders() {
 	s.encoderCreated.Inc(1)
 }
+
+// WriteType is an enum for warm/cold write types.
+type WriteType int
+
+const (
+	// UndefinedWriteType is an undefined write type.
+	UndefinedWriteType WriteType = iota
+
+	// WarmWrite represents warm writes (within the buffer past/future window).
+	WarmWrite
+
+	// ColdWrite represents cold writes (outside the buffer past/future window).
+	ColdWrite
+)
+
+type WriteOptions struct {
+	writeType WriteType
+}
+
+func (o *WriteOptions) SetWriteType(value WriteType) *WriteOptions {
+	opts := *o
+	opts.writeType = value
+	return &opts
+}
+
+func (o *WriteOptions) WriteType() WriteType {
+	return o.writeType
+}
+
+func NewWriteOptions() *WriteOptions {
+	return &WriteOptions{}
+}
```

Comment:
Can we get rid of these setter/getter methods?

Also we should remove `NewWriteOptions()` and change the types everywhere to take just a struct by value `WriteOptions` instead of a pointer `*WriteOptions`.

If it's a pointer type then we need to allocate on the heap, but since it's not stateful we should just copy and pass it everywhere by the stack (much cheaper in terms of memory management).

i.e. remove `NewWriteOptions(...)` and the getters/setters and change all types from:
```go
func DoThing(arg string, opts *WriteOptions) { ... }
```

To:
```go
func DoThing(arg string, opts WriteOptions) { ... }
```

Reply:
Ah, I see. thanks for the explanation, it makes sense. yes, if it is pass over the stack, then immutability is not needed. I did not really get what Richard is trying to say the first time.